### PR TITLE
feat: Add tooling to generate stubs for the Public PHP API

### PR DESCRIPTION
**Pull request by PROFeNoM**
### Description

[AIDM-460](https://datadoghq.atlassian.net/browse/AIDM-460) & [AIDM-463](https://datadoghq.atlassian.net/browse/AIDM-463)

This PR adds simple tooling to add one stub file (`src/ddtrace_php_api.stubs.php`) for all the Public PHP API files (pulled from the files declared in the bridge).

This allows us to generate a stub file locally using `make generate_stubs` (PHP 8+).

Additionally, all stubs files (there are four) will be merged together and added to the final artifact as `datadog-tracer.stubs.php`.

The `generated files up-to-date` job will regenerate PHP stubs and compare them on each commit, failing if there is a difference.

### Reviewer checklist
- [ ] Test coverage seems ok.
- [ ] Appropriate labels assigned.

[AIDM-460]: https://datadoghq.atlassian.net/browse/AIDM-460?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[AIDM-463]: https://datadoghq.atlassian.net/browse/AIDM-463?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: tooling/bin/generate-final-artifact.sh [hunk 1]
```diff
@@ -152,6 +152,33 @@ for architecture in "${architectures[@]}"; do
...
     fi
 
+    ########################
+    # PHP Stubs
+    ########################
+    stubs=(
+        "/home/circleci/app/src/ddtrace_php_api.stubs.php"
+        "/home/circleci/app/ext/ddtrace.stub.php"
+        "/home/circleci/app/ext/hook/uhook.stub.php"
+        "/home/circleci/app/ext/hook/uhook_attributes.stub.php"
```
Comment (PROFeNoM) [inline tooling/bin/generate-final-artifact.sh:162]:
```suggestion
        "/home/circleci/datadog/src/ddtrace_php_api.stubs.php"
        "/home/circleci/datadog/ext/ddtrace.stub.php"
        "/home/circleci/datadog/ext/hook/uhook.stub.php"
        "/home/circleci/datadog/ext/hook/uhook_attributes.stub.php"
```

File: tooling/bin/generate-final-artifact.sh [hunk 2]
```diff
@@ -152,6 +152,33 @@ for architecture in "${architectures[@]}"; do
...
+    stubs=(
+        "/home/circleci/datadog/src/ddtrace_php_api.stubs.php"
+        "/home/circleci/datadog/ext/ddtrace.stub.php"
+        "/home/circleci/datadog/ext/hook/uhook.stub.php"
+        "/home/circleci/datadog/ext/hook/uhook_attributes.stub.php"
+    )
+
+    mergedStubs=""
+    for stub in "${stubs[@]}"; do
+        content=$(<"$stub")
```
Comment (bwoebi) [inline tooling/bin/generate-final-artifact.sh:167]: What's that? The same than `content=$stub`?

File: tooling/bin/generate-final-artifact.sh [hunk 3]
```diff
@@ -152,6 +152,33 @@ for architecture in "${architectures[@]}"; do
...
     fi
 
+    ########################
+    # PHP Stubs
+    ########################
+    stubs=(
+        "/home/circleci/datadog/src/ddtrace_php_api.stubs.php"
+        "/home/circleci/datadog/ext/ddtrace.stub.php"
+        "/home/circleci/datadog/ext/hook/uhook.stub.php"
+        "/home/circleci/datadog/ext/hook/uhook_attributes.stub.php"
```
Comment (bwoebi) [inline tooling/bin/generate-final-artifact.sh:162]: Could you please pass `$(PROJECT_ROOT)` in the Makefile and replace /home/circleci/datadog by $3?

**Comment by bwoebi** [comment]
Could you please rename the `cbindgen up-to-date` step in circleci to `generated files up-to-date` and add `make generate_stubs; git diff --exit-code src/ddtrace_php_api.stubs.php`? (and switch the container to a buster container which also includes php)

**Comment by bwoebi** [comment]
Also, can you please put a copy of the generated merged stub file in the $packages_build_dir directly? That way it'll also appear in the artifacts directly, for easy download directly from the artifacts page.

File: tooling/bin/generate-final-artifact.sh (hunk 2 above)
Comment (PROFeNoM) [reply tooling/bin/generate-final-artifact.sh:167]: `$(<"$stub")` reads the contents of the file specified by "$stub" into the variable `content`. `stub` here will be one of the following:

```bash
(
        "/home/circleci/app/src/ddtrace_php_api.stubs.php"
        "/home/circleci/app/ext/ddtrace.stub.php"
        "/home/circleci/app/ext/hook/uhook.stub.php"
        "/home/circleci/app/ext/hook/uhook_attributes.stub.php"
)
```

I really am not a bash expert ^^ but I believe this is right 🤔 

I read the content of the stub file, remove the php handle, and then add the content to the global merged stubs.

Comment (bwoebi) [reply tooling/bin/generate-final-artifact.sh:167]: oh you are right :-) Never seen that specific syntax soup. I thought that was where the good old `cat` was needed, so disregard that.

**Comment by bwoebi** [comment]
```
php -l datadog-tracer.stubs.php 
Parse error: syntax error, unexpected character 0x1B, expecting "{" in datadog-tracer.stubs.php on line 1484
```

Somehow the string sequence `\e` got re-interpreted as a literal escape character in the output.

File: tooling/bin/generate-final-artifact.sh [hunk 4]
```diff
@@ -152,6 +152,26 @@ for architecture in "${architectures[@]}"; do
...
+
+    mergedStubs=""
+    for stub in "${stubs[@]}"; do
+        content=$(<"$stub")
+        content="${content#<?php}"
+        mergedStubs+="$content"
+    done
+
+    stub="<?php\n$mergedStubs"
+    echo -e "$stub" > "$packages_build_dir/datadog-tracer.stubs.php"
```
Comment (bwoebi) [inline tooling/bin/generate-final-artifact.sh:173]: I think it's this echo `-e` which causes that.
```suggestion
    stub=$'<?php\n'"$mergedStubs"
    echo "$stub" > "$packages_build_dir/datadog-tracer.stubs.php"
```

Not sure, but does this work?

**Review by bwoebi — APPROVED** [review]
Nice! Thanks Alexandre :-)

**Pull request merged**